recipes/full_finetune_distributed.py

```diff
@@ -697,15 +703,31 @@ def train(self) -> None:
                 # Compute loss
                 # Loss is normalized by default so we multiply by the number of tokens
                 # This way we can normalize by the total number of tokens if we're accumulating gradients
-                running_loss += self._loss_fn(logits, labels) * current_num_tokens
+                current_loss = self._loss_fn(logits, labels) * current_num_tokens
 
                 # free logits otherwise it peaks backward memory
                 del logits
 
+                running_loss += current_loss
+
+                # For optimizer in backward, we need to normalize before calling backward
+                # This case and gradient accumulation are mutually exclusive
+                if self._optimizer_in_bwd:
+                    torch.distributed.all_reduce(num_tokens)
+                    torch.distributed.all_reduce(running_loss)
+                    current_loss = current_loss / num_tokens
+
+                current_loss.backward()
+
                 # Step with optimizer
                 if (idx + 1) % self._gradient_accumulation_steps == 0:
-                    loss = running_loss / num_tokens
-                    loss.backward()
+                    if not self._optimizer_in_bwd:
+                        # Get total number of tokens across all ranks to normalize gradients
+                        torch.distributed.all_reduce(num_tokens)
+                        # This will ensure that the logged loss matches what we're optimizing
+                        torch.distributed.all_reduce(running_loss)
+                        # Manually scale the gradients from unnormalized loss by total # of tokens
+                        training.scale_grads(self._model, 1 / num_tokens)
```

Comment:
1) The way num_tokens is being calculated/reset seems a bit fishy. 
Lets say you have 2 gpus, and curr_num_tokens = [5,20] in gpus 0,1 for the first step, and [10, 3] for the second step.

For gpu 0, step 1:
num_tokens += 5
For gpu 1, step 1
num_tokens += 20

then you do all reduce, so both should have 25

You did NOT reset num_tokens, since we are accumulating gradients,

so on step 2, num_tokens is:
For gpu 0, step 2:
num_tokens =  25+10
For gpu 1, step 1
num_tokens = 25+3

you do all_reduce, and its
35+28.

Isnt this whats happening?

Then you do all reduce again when accumulating gradients. Does it make num_tokens * 2?

Reply:
2) Does all reduce require synchronization? If so, i wonder if we can make it async, and start it before running the model, so it has enough time. Maybe not for this PR..

Reply:
all_reduce on tokens only happens just before step (but lmk if i'm misunderstanding your comment)